metadata_ui.js

```diff
@@ -0,0 +1,194 @@
+        if (typeof opts.args[1] === 'object') {
+          if (opts.args[1].metadata && typeof opts.args[1] === 'object'
+            && opts.args[1].tests && typeof opts.args[1].tests === 'function') {
+            metadata = opts.args[1].metadata;
+            fn = opts.args[1].tests;
+          } else {
+            throw new Error('If passing an object as the second parameter, it must be of the form { metadata: <obj>, tests: <fn> }');
+          }
+        } else if (typeof opts.args[1] === 'function') {
+          fn = opts.args[1];
+        } else {
+          throw new Error('Incorrect suite usage. Must be either "describe(<string>, { <object>, <function> })"" or "describe(<string>, <function)"');
+        }
+      } else if (opts.args.length === 3) { // Metadata as a param: describe(title, meta, fn)
+        if (opts.args[1] && typeof opts.args[1] === 'object'
+          && opts.args[2] && typeof opts.args[2] === 'function') {
+          metadata = opts.args[1];
+          fn = opts.args[2];
+        } else {
+          throw new Error('If passing three parameters, they must be of the form "describe(<string>, <object>, <function>)"');
+        }
+      } else if (opts.args.length > 3) {
+        throw new Error('Too many arguments passed.');
+      }
+
+      // Creating the Suite object
+      var newSuite = Suite.create(suites[0], title);
+      newSuite.pending = Boolean(opts.pending);
+      newSuite.file = file;
+      suites.unshift(newSuite);
+      if (opts.isOnly) {
+        newSuite.parent._onlySuites = newSuite.parent._onlySuites.concat(newSuite);
+        mocha.options.hasOnly = true;
+      }
+      newSuite.metadata = metadata || {};
+      if (typeof fn === 'function') {
+        fn.call(newSuite);
+        suites.shift();
+      } else if (typeof fn === 'undefined' && !newSuite.pending) {
+        throw new Error('Suite "' + newSuite.fullTitle() + '" was defined but no callback was supplied. Supply a callback or explicitly skip the suite.');
+      }
+
+      return newSuite;
+    };
+
+    /**
+     * Describe a "suite" with the given `title`
+     * and callback `fn` containing nested suites
+     * and/or tests.
+     */
+
+    context.describe = context.context = function() {
+      return _create({
+        args: arguments
+      });
+    };
+
+    /**
+     * Pending describe.
+     */
+
+    context.xdescribe = context.xcontext = context.describe.skip = function(title, fn) {
+      return _create({
+        args: arguments,
+        pending: true
+      });
+    };
+
+    /**
+     * Exclusive suite.
+     */
+
+    context.describe.only = function(title, fn) {
+      return _create({
+        args: arguments,
+        isOnly: true
+      });
+    };
+
+    /**
+     * Describe a specification or test-case
+     * with the given `title` and callback `fn`
+     * acting as a thunk.
+     */
+
+    context.it = context.specify = function() {
+      var title, metadata, fn;
+
+      if (arguments.length < 2) {
```

Comment:
the arguments parsing here looks a lot like the parsing for the suite-level above in `_create`. Is it possible to refactor them to share the same code?